### PR TITLE
coord: Make it possible to dump state for debugging

### PR DESCRIPTION
**Pull request by ParkMyCar**
Follow up to an incident, it would be nice to be able to dump some of the Coordinator's state so we can determine what read holds are currently outstanding. This PR adds a new endpoint to the internal HTTP server `/api/coordinator/dump` that will return part of the Coordinator's state as a JSON object.

It also updates the read holds consistency check to make sure the connection exists for all txn related read holds.

### Motivation

Debugging tool

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - N/a

### CODE REVIEW

**Review by mjibson — APPROVED** [review]

**Comment by teskje** [comment]
This is great! Once this is merged I'd love to send a follow-up that also includes (some of) the controller state in this dump.

How did you select JSON as the output format, and which fields to include? My instinct would have been to implement `Debug` on the controller and include everything that's reasonably debug-printable, to ensure we don't miss any data we could have included when we debug an incident.

**Review by ggevay — COMMENTED** [review]
Good idea!

**Comment by aalexandrov** [comment]
> How did you select JSON as the output format, and which fields to include?

Not having a bit stake in the JSON vs Debug decision here, but I want to point out that JSON would allow us to write tooling to automate answering certain types of common questions with `jq` (like the ones @mgree is proposing for query plan analysis in #25972).

File: src/adapter/src/coord.rs [hunk 1]
```diff
@@ -2699,6 +2702,92 @@ impl Coordinator {
...
+            (
+                "pending_peeks".to_string(),
+                serde_json::to_value(pending_peeks)?,
+            ),
+            (
+                "client_pending_peeks".to_string(),
+                serde_json::to_value(client_pending_peeks)?,
+            ),
+            (
+                "pending_real_time_recenty_timestamp".to_string(),
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/adapter/src/coord.rs]: Thanks!

**Comment by ParkMyCar** [comment]
> This is great! Once this is merged I'd love to send a follow-up that also includes (some of) the controller state in this dump.

That sounds great!

> How did you select JSON as the output format, and which fields to include? My instinct would have been to implement Debug on the controller and include everything that's reasonably debug-printable, to ensure we don't miss any data we could have included when we debug an incident.

The first version of this that I implemented just `Debug` printed the entire Coordinator and wrote that out to traces. There were two issues with this though:
1. _Way_ too much data. The output was almost useless because the trace was so large and `Debug` formatting isn't JSON compatible so it was so hard to use.
2. Security. Printing the `Debug` format of the state seemed delicate in case down the road there is some info that we wouldn't want ending up persisted anywhere.

To me the next logical thing was JSON formatting since it's:
1. Separate from the `Debug` impl so we can include only the fields we want.
2. Good tools exists to parse this, e.g. `jq` or your browser, so working with the output and ignoring sections of it would be easy.

**Comment by teskje** [comment]
That sounds reasonable to me!

**Pull request merged**